// @metamask/snaps-controllers/dist/cjs/snaps/endowments/home-page.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "homePageEndowmentBuilder", {
    enumerable: true,
    get: function() {
        return homePageEndowmentBuilder;
    }
});
const _permissioncontroller = require("@metamask/permission-controller");
const _enum = require("./enum");
const permissionName = _enum.SnapEndowments.HomePage;
/**
 * `endowment:page-home` returns nothing; it is intended to be used as a
 * flag by the snap controller to detect whether the snap has the capability to
 * use the snap home page feature.
 *
 * @param _builderOptions - Optional specification builder options.
 * @returns The specification for the `snap-pages` endowment.
 */ const specificationBuilder = (_builderOptions)=>{
    return {
        permissionType: _permissioncontroller.PermissionType.Endowment,
        targetName: permissionName,
        allowedCaveats: null,
        endowmentGetter: (_getterOptions)=>undefined,
        subjectTypes: [
            _permissioncontroller.SubjectType.Snap
        ]
    };
};
const homePageEndowmentBuilder = Object.freeze({
    targetName: permissionName,
    specificationBuilder
});
