// @metamask/controller-utils/dist/util.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});



























var _chunkEDY72VWLjs = require('./chunk-EDY72VWL.js');
require('./chunk-X4PYR7CK.js');
require('./chunk-LGGBBU6G.js');




























exports.BNToHex = _chunkEDY72VWLjs.BNToHex; exports.PROTOTYPE_POLLUTION_BLOCKLIST = _chunkEDY72VWLjs.PROTOTYPE_POLLUTION_BLOCKLIST; exports.convertHexToDecimal = _chunkEDY72VWLjs.convertHexToDecimal; exports.fetchWithErrorHandling = _chunkEDY72VWLjs.fetchWithErrorHandling; exports.fractionBN = _chunkEDY72VWLjs.fractionBN; exports.fromHex = _chunkEDY72VWLjs.fromHex; exports.getBuyURL = _chunkEDY72VWLjs.getBuyURL; exports.gweiDecToWEIBN = _chunkEDY72VWLjs.gweiDecToWEIBN; exports.handleFetch = _chunkEDY72VWLjs.handleFetch; exports.hexToBN = _chunkEDY72VWLjs.hexToBN; exports.hexToText = _chunkEDY72VWLjs.hexToText; exports.isNonEmptyArray = _chunkEDY72VWLjs.isNonEmptyArray; exports.isPlainObject = _chunkEDY72VWLjs.isPlainObject; exports.isSafeChainId = _chunkEDY72VWLjs.isSafeChainId; exports.isSafeDynamicKey = _chunkEDY72VWLjs.isSafeDynamicKey; exports.isSmartContractCode = _chunkEDY72VWLjs.isSmartContractCode; exports.isValidHexAddress = _chunkEDY72VWLjs.isValidHexAddress; exports.isValidJson = _chunkEDY72VWLjs.isValidJson; exports.normalizeEnsName = _chunkEDY72VWLjs.normalizeEnsName; exports.query = _chunkEDY72VWLjs.query; exports.safelyExecute = _chunkEDY72VWLjs.safelyExecute; exports.safelyExecuteWithTimeout = _chunkEDY72VWLjs.safelyExecuteWithTimeout; exports.successfulFetch = _chunkEDY72VWLjs.successfulFetch; exports.timeoutFetch = _chunkEDY72VWLjs.timeoutFetch; exports.toChecksumHexAddress = _chunkEDY72VWLjs.toChecksumHexAddress; exports.toHex = _chunkEDY72VWLjs.toHex; exports.weiHexToGweiDec = _chunkEDY72VWLjs.weiHexToGweiDec;
