// @metamask/snaps-controllers/dist/cjs/services/node/NodeThreadExecutionService.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "NodeThreadExecutionService", {
    enumerable: true,
    get: function() {
        return NodeThreadExecutionService;
    }
});
const _postmessagestream = require("@metamask/post-message-stream");
const _worker_threads = require("worker_threads");
const _ = require("..");
class NodeThreadExecutionService extends _.AbstractExecutionService {
    async initEnvStream() {
        const worker = new _worker_threads.Worker(require.resolve('@metamask/snaps-execution-environments/dist/browserify/node-thread/bundle.js'), {
            stdout: true,
            stderr: true
        });
        // Capturing `stdout` and `stderr` from the worker prevents the worker from
        // writing to them directly, making it easier to capture them Jest.
        worker.stdout.on('data', (data)=>{
            // eslint-disable-next-line no-console
            console.log(data.toString());
        });
        worker.stderr.on('data', (data)=>{
            // eslint-disable-next-line no-console
            console.error(data.toString());
        });
        const stream = new _postmessagestream.ThreadParentMessageStream({
            thread: worker
        });
        return Promise.resolve({
            worker,
            stream
        });
    }
    async terminateJob(jobWrapper) {
        await jobWrapper.worker.terminate();
    }
}
