// @metamask/permission-controller/dist/chunk-YRNH4R3G.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkU5LTEXSUjs = require('./chunk-U5LTEXSU.js');

// src/rpc-methods/requestPermissions.ts
var _controllerutils = require('@metamask/controller-utils');
var requestPermissionsHandler = {
  methodNames: ["wallet_requestPermissions" /* requestPermissions */],
  implementation: requestPermissionsImplementation,
  hookNames: {
    requestPermissionsForOrigin: true
  }
};
async function requestPermissionsImplementation(req, res, _next, end, { requestPermissionsForOrigin }) {
  const { params } = req;
  if (!Array.isArray(params) || !_controllerutils.isPlainObject.call(void 0, params[0])) {
    return end(_chunkU5LTEXSUjs.invalidParams.call(void 0, { data: { request: req } }));
  }
  const [requestedPermissions] = params;
  const [grantedPermissions] = await requestPermissionsForOrigin(
    requestedPermissions
  );
  res.result = Object.values(grantedPermissions);
  return end();
}



exports.requestPermissionsHandler = requestPermissionsHandler;
