// @metamask/permission-controller/dist/errors.mjs
import {
  CaveatAlreadyExistsError,
  CaveatDoesNotExistError,
  CaveatInvalidJsonError,
  CaveatMissingValueError,
  CaveatSpecificationMismatchError,
  DuplicateCaveatError,
  EndowmentPermissionDoesNotExistError,
  ForbiddenCaveatError,
  InvalidApprovedPermissionError,
  InvalidCaveatError,
  InvalidCaveatFieldsError,
  InvalidCaveatTypeError,
  InvalidCaveatsPropertyError,
  InvalidSubjectIdentifierError,
  PermissionDoesNotExistError,
  PermissionsRequestNotFoundError,
  UnrecognizedCaveatTypeError,
  UnrecognizedSubjectError,
  internalError,
  invalidParams,
  methodNotFound,
  unauthorized,
  userRejectedRequest
} from "./chunk-FEXS6SJF.mjs";
import "./chunk-ZSTKOCBT.mjs";
export {
  CaveatAlreadyExistsError,
  CaveatDoesNotExistError,
  CaveatInvalidJsonError,
  CaveatMissingValueError,
  CaveatSpecificationMismatchError,
  DuplicateCaveatError,
  EndowmentPermissionDoesNotExistError,
  ForbiddenCaveatError,
  InvalidApprovedPermissionError,
  InvalidCaveatError,
  InvalidCaveatFieldsError,
  InvalidCaveatTypeError,
  InvalidCaveatsPropertyError,
  InvalidSubjectIdentifierError,
  PermissionDoesNotExistError,
  PermissionsRequestNotFoundError,
  UnrecognizedCaveatTypeError,
  UnrecognizedSubjectError,
  internalError,
  invalidParams,
  methodNotFound,
  unauthorized,
  userRejectedRequest
};
