// @metamask/permission-controller/dist/index.mjs
import {
  rpc_methods_exports
} from "./chunk-5C5SZO74.mjs";
import "./chunk-73M67Z5X.mjs";
import "./chunk-YVXIWLBE.mjs";
import "./chunk-BEQEQLRE.mjs";
import {
  CaveatMutatorOperation,
  PermissionController
} from "./chunk-VFQJB7BG.mjs";
import {
  decorateWithCaveats,
  isRestrictedMethodCaveatSpecification
} from "./chunk-XBFHEZRU.mjs";
import {
  PermissionType,
  constructPermission,
  findCaveat,
  hasSpecificationType
} from "./chunk-ODCVB4BB.mjs";
import {
  SubjectMetadataController,
  SubjectType
} from "./chunk-4D2RVWNF.mjs";
import "./chunk-N4KQ2BHF.mjs";
import {
  CaveatAlreadyExistsError,
  CaveatDoesNotExistError,
  CaveatInvalidJsonError,
  CaveatMissingValueError,
  CaveatSpecificationMismatchError,
  DuplicateCaveatError,
  EndowmentPermissionDoesNotExistError,
  ForbiddenCaveatError,
  InvalidApprovedPermissionError,
  InvalidCaveatError,
  InvalidCaveatFieldsError,
  InvalidCaveatTypeError,
  InvalidCaveatsPropertyError,
  InvalidSubjectIdentifierError,
  PermissionDoesNotExistError,
  PermissionsRequestNotFoundError,
  UnrecognizedCaveatTypeError,
  UnrecognizedSubjectError,
  internalError,
  invalidParams,
  methodNotFound,
  unauthorized,
  userRejectedRequest
} from "./chunk-FEXS6SJF.mjs";
import {
  MethodNames
} from "./chunk-KVKKNDYV.mjs";
import "./chunk-ZSTKOCBT.mjs";
export {
  CaveatAlreadyExistsError,
  CaveatDoesNotExistError,
  CaveatInvalidJsonError,
  CaveatMissingValueError,
  CaveatMutatorOperation,
  CaveatSpecificationMismatchError,
  DuplicateCaveatError,
  EndowmentPermissionDoesNotExistError,
  ForbiddenCaveatError,
  InvalidApprovedPermissionError,
  InvalidCaveatError,
  InvalidCaveatFieldsError,
  InvalidCaveatTypeError,
  InvalidCaveatsPropertyError,
  InvalidSubjectIdentifierError,
  MethodNames,
  PermissionController,
  PermissionDoesNotExistError,
  PermissionType,
  PermissionsRequestNotFoundError,
  SubjectMetadataController,
  SubjectType,
  UnrecognizedCaveatTypeError,
  UnrecognizedSubjectError,
  constructPermission,
  decorateWithCaveats,
  findCaveat,
  hasSpecificationType,
  internalError,
  invalidParams,
  isRestrictedMethodCaveatSpecification,
  methodNotFound,
  rpc_methods_exports as permissionRpcMethods,
  unauthorized,
  userRejectedRequest
};
