// @metamask/snaps-utils/dist/cjs/eval-worker.js
// eslint-disable-next-line import/no-unassigned-import
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
require("ses/lockdown");
const _fs = require("fs");
const _handlertypes = require("./handler-types");
const _mock = require("./mock");
lockdown({
    consoleTaming: 'unsafe',
    errorTaming: 'unsafe',
    mathTaming: 'unsafe',
    dateTaming: 'unsafe',
    overrideTaming: 'severe',
    // We disable domain taming, because it does not work in certain cases when
    // running tests. This is unlikely to be a problem in production, because
    // Node.js domains are deprecated.
    domainTaming: 'unsafe'
});
const snapFilePath = process.argv[2];
const snapModule = {
    exports: {}
};
const compartment = new Compartment({
    ...(0, _mock.generateMockEndowments)(),
    module: snapModule,
    exports: snapModule.exports
});
// Mirror BaseSnapExecutor
compartment.globalThis.self = compartment.globalThis;
compartment.globalThis.global = compartment.globalThis;
compartment.globalThis.window = compartment.globalThis;
compartment.evaluate((0, _fs.readFileSync)(snapFilePath, 'utf8'));
const invalidExports = Object.keys(snapModule.exports).filter((snapExport)=>!_handlertypes.SNAP_EXPORT_NAMES.includes(snapExport));
if (invalidExports.length > 0) {
    // eslint-disable-next-line no-console
    console.warn(`Invalid snap exports detected:\n${invalidExports.join('\n')}`);
}
setTimeout(()=>process.exit(0), 1000); // Hack to ensure worker exits
