// @metamask/snaps-rpc-methods/dist/cjs/restricted/dialog.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    dialogBuilder: function() {
        return dialogBuilder;
    },
    getDialogImplementation: function() {
        return getDialogImplementation;
    }
});
const _permissioncontroller = require("@metamask/permission-controller");
const _rpcerrors = require("@metamask/rpc-errors");
const _snapssdk = require("@metamask/snaps-sdk");
const _snapsutils = require("@metamask/snaps-utils");
const _superstruct = require("superstruct");
const methodName = 'snap_dialog';
const PlaceholderStruct = (0, _superstruct.optional)((0, _superstruct.size)((0, _superstruct.string)(), 1, 40));
/**
 * The specification builder for the `snap_dialog` permission. `snap_dialog`
 * lets the Snap display one of the following dialogs to the user:
 * - An alert, for displaying information.
 * - A confirmation, for accepting or rejecting some action.
 * - A prompt, for inputting some information.
 *
 * @param options - The specification builder options.
 * @param options.allowedCaveats - The optional allowed caveats for the
 * permission.
 * @param options.methodHooks - The RPC method hooks needed by the method
 * implementation.
 * @returns The specification for the `snap_dialog` permission.
 */ const specificationBuilder = ({ allowedCaveats = null, methodHooks })=>{
    return {
        permissionType: _permissioncontroller.PermissionType.RestrictedMethod,
        targetName: methodName,
        allowedCaveats,
        methodImplementation: getDialogImplementation(methodHooks),
        subjectTypes: [
            _permissioncontroller.SubjectType.Snap
        ]
    };
};
const methodHooks = {
    showDialog: true,
    isOnPhishingList: true,
    maybeUpdatePhishingList: true
};
const dialogBuilder = Object.freeze({
    targetName: methodName,
    specificationBuilder,
    methodHooks
});
// Note: We use `type` here instead of `object` because `type` does not validate
// the keys of the object, which is what we want.
const BaseParamsStruct = (0, _superstruct.type)({
    type: (0, _superstruct.enums)([
        _snapssdk.DialogType.Alert,
        _snapssdk.DialogType.Confirmation,
        _snapssdk.DialogType.Prompt
    ])
});
const AlertParametersStruct = (0, _superstruct.object)({
    type: (0, _snapssdk.enumValue)(_snapssdk.DialogType.Alert),
    content: _snapssdk.ComponentStruct
});
const ConfirmationParametersStruct = (0, _superstruct.object)({
    type: (0, _snapssdk.enumValue)(_snapssdk.DialogType.Confirmation),
    content: _snapssdk.ComponentStruct
});
const PromptParametersStruct = (0, _superstruct.object)({
    type: (0, _snapssdk.enumValue)(_snapssdk.DialogType.Prompt),
    content: _snapssdk.ComponentStruct,
    placeholder: PlaceholderStruct
});
const DialogParametersStruct = (0, _superstruct.union)([
    AlertParametersStruct,
    ConfirmationParametersStruct,
    PromptParametersStruct
]);
const structs = {
    [_snapssdk.DialogType.Alert]: AlertParametersStruct,
    [_snapssdk.DialogType.Confirmation]: ConfirmationParametersStruct,
    [_snapssdk.DialogType.Prompt]: PromptParametersStruct
};
function getDialogImplementation({ showDialog, isOnPhishingList, maybeUpdatePhishingList }) {
    return async function dialogImplementation(args) {
        const { params, context: { origin } } = args;
        const validatedType = getValidatedType(params);
        const validatedParams = getValidatedParams(params, structs[validatedType]);
        const { content } = validatedParams;
        await maybeUpdatePhishingList();
        (0, _snapsutils.validateComponentLinks)(content, isOnPhishingList);
        const placeholder = validatedParams.type === _snapssdk.DialogType.Prompt ? validatedParams.placeholder : undefined;
        return showDialog(origin, validatedType, content, placeholder);
    };
}
/**
 * Get the validated type of the dialog parameters. Throws an error if the type
 * is invalid.
 *
 * @param params - The parameters to validate.
 * @returns The validated type of the dialog parameters.
 */ function getValidatedType(params) {
    try {
        return (0, _superstruct.create)(params, BaseParamsStruct).type;
    } catch (error) {
        throw _rpcerrors.rpcErrors.invalidParams({
            message: `The "type" property must be one of: ${Object.values(_snapssdk.DialogType).join(', ')}.`
        });
    }
}
/**
 * Validates the confirm method `params` and returns them cast to the correct
 * type. Throws if validation fails.
 *
 * @param params - The unvalidated params object from the method request.
 * @param struct - The struct to validate the params against.
 * @returns The validated confirm method parameter object.
 */ function getValidatedParams(params, struct) {
    try {
        return (0, _superstruct.create)(params, struct);
    } catch (error) {
        if (error instanceof _superstruct.StructError) {
            const { key, type: errorType } = error;
            if (key === 'placeholder' && errorType === 'never') {
                throw _rpcerrors.rpcErrors.invalidParams({
                    message: 'Invalid params: Alerts or confirmations may not specify a "placeholder" field.'
                });
            }
            throw _rpcerrors.rpcErrors.invalidParams({
                message: `Invalid params: ${error.message}.`
            });
        }
        /* istanbul ignore next */ throw _rpcerrors.rpcErrors.internal();
    }
}
