// @metamask/snaps-rpc-methods/dist/cjs/restricted/notify.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    specificationBuilder: function() {
        return specificationBuilder;
    },
    notifyBuilder: function() {
        return notifyBuilder;
    },
    getImplementation: function() {
        return getImplementation;
    },
    getValidatedParams: function() {
        return getValidatedParams;
    }
});
const _permissioncontroller = require("@metamask/permission-controller");
const _rpcerrors = require("@metamask/rpc-errors");
const _snapssdk = require("@metamask/snaps-sdk");
const _snapsutils = require("@metamask/snaps-utils");
const _utils = require("@metamask/utils");
const methodName = 'snap_notify';
const specificationBuilder = ({ allowedCaveats = null, methodHooks })=>{
    return {
        permissionType: _permissioncontroller.PermissionType.RestrictedMethod,
        targetName: methodName,
        allowedCaveats,
        methodImplementation: getImplementation(methodHooks),
        subjectTypes: [
            _permissioncontroller.SubjectType.Snap
        ]
    };
};
const methodHooks = {
    showNativeNotification: true,
    showInAppNotification: true,
    isOnPhishingList: true,
    maybeUpdatePhishingList: true
};
const notifyBuilder = Object.freeze({
    targetName: methodName,
    specificationBuilder,
    methodHooks
});
function getImplementation({ showNativeNotification, showInAppNotification, isOnPhishingList, maybeUpdatePhishingList }) {
    return async function implementation(args) {
        const { params, context: { origin } } = args;
        const validatedParams = getValidatedParams(params);
        await maybeUpdatePhishingList();
        (0, _snapsutils.validateTextLinks)(validatedParams.message, isOnPhishingList);
        switch(validatedParams.type){
            case _snapssdk.NotificationType.Native:
                return await showNativeNotification(origin, validatedParams);
            case _snapssdk.NotificationType.InApp:
                return await showInAppNotification(origin, validatedParams);
            default:
                throw _rpcerrors.rpcErrors.invalidParams({
                    message: 'Must specify a valid notification "type".'
                });
        }
    };
}
function getValidatedParams(params) {
    if (!(0, _utils.isObject)(params)) {
        throw _rpcerrors.rpcErrors.invalidParams({
            message: 'Expected params to be a single object.'
        });
    }
    const { type, message } = params;
    if (!type || typeof type !== 'string' || !Object.values(_snapssdk.NotificationType).includes(type)) {
        throw _rpcerrors.rpcErrors.invalidParams({
            message: 'Must specify a valid notification "type".'
        });
    }
    // Set to the max message length on a Mac notification for now.
    if (!message || typeof message !== 'string' || message.length >= 50) {
        throw _rpcerrors.rpcErrors.invalidParams({
            message: 'Must specify a non-empty string "message" less than 50 characters long.'
        });
    }
    return params;
}
