// @metamask/controller-utils/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});



var _chunkM6G3ISXGjs = require('./chunk-M6G3ISXG.js');
require('./chunk-DBJSLKNR.js');



























var _chunkEDY72VWLjs = require('./chunk-EDY72VWL.js');
























var _chunkX4PYR7CKjs = require('./chunk-X4PYR7CK.js');








var _chunkLGGBBU6Gjs = require('./chunk-LGGBBU6G.js');




























































exports.ASSET_TYPES = _chunkX4PYR7CKjs.ASSET_TYPES; exports.ApprovalType = _chunkX4PYR7CKjs.ApprovalType; exports.BNToHex = _chunkEDY72VWLjs.BNToHex; exports.BUILT_IN_NETWORKS = _chunkX4PYR7CKjs.BUILT_IN_NETWORKS; exports.BuiltInNetworkName = _chunkLGGBBU6Gjs.BuiltInNetworkName; exports.CHAIN_ID_TO_ETHERS_NETWORK_NAME_MAP = _chunkX4PYR7CKjs.CHAIN_ID_TO_ETHERS_NETWORK_NAME_MAP; exports.ChainId = _chunkLGGBBU6Gjs.ChainId; exports.ERC1155 = _chunkX4PYR7CKjs.ERC1155; exports.ERC1155_INTERFACE_ID = _chunkX4PYR7CKjs.ERC1155_INTERFACE_ID; exports.ERC1155_METADATA_URI_INTERFACE_ID = _chunkX4PYR7CKjs.ERC1155_METADATA_URI_INTERFACE_ID; exports.ERC1155_TOKEN_RECEIVER_INTERFACE_ID = _chunkX4PYR7CKjs.ERC1155_TOKEN_RECEIVER_INTERFACE_ID; exports.ERC20 = _chunkX4PYR7CKjs.ERC20; exports.ERC721 = _chunkX4PYR7CKjs.ERC721; exports.ERC721_ENUMERABLE_INTERFACE_ID = _chunkX4PYR7CKjs.ERC721_ENUMERABLE_INTERFACE_ID; exports.ERC721_INTERFACE_ID = _chunkX4PYR7CKjs.ERC721_INTERFACE_ID; exports.ERC721_METADATA_INTERFACE_ID = _chunkX4PYR7CKjs.ERC721_METADATA_INTERFACE_ID; exports.FALL_BACK_VS_CURRENCY = _chunkX4PYR7CKjs.FALL_BACK_VS_CURRENCY; exports.GANACHE_CHAIN_ID = _chunkX4PYR7CKjs.GANACHE_CHAIN_ID; exports.GWEI = _chunkX4PYR7CKjs.GWEI; exports.IPFS_DEFAULT_GATEWAY_URL = _chunkX4PYR7CKjs.IPFS_DEFAULT_GATEWAY_URL; exports.InfuraNetworkType = _chunkLGGBBU6Gjs.InfuraNetworkType; exports.MAX_SAFE_CHAIN_ID = _chunkX4PYR7CKjs.MAX_SAFE_CHAIN_ID; exports.NFT_API_BASE_URL = _chunkX4PYR7CKjs.NFT_API_BASE_URL; exports.NetworkType = _chunkLGGBBU6Gjs.NetworkType; exports.NetworksTicker = _chunkLGGBBU6Gjs.NetworksTicker; exports.OPENSEA_PROXY_URL = _chunkX4PYR7CKjs.OPENSEA_PROXY_URL; exports.ORIGIN_METAMASK = _chunkX4PYR7CKjs.ORIGIN_METAMASK; exports.RPC = _chunkX4PYR7CKjs.RPC; exports.TESTNET_TICKER_SYMBOLS = _chunkX4PYR7CKjs.TESTNET_TICKER_SYMBOLS; exports.convertHexToDecimal = _chunkEDY72VWLjs.convertHexToDecimal; exports.detectSIWE = _chunkM6G3ISXGjs.detectSIWE; exports.fetchWithErrorHandling = _chunkEDY72VWLjs.fetchWithErrorHandling; exports.fractionBN = _chunkEDY72VWLjs.fractionBN; exports.fromHex = _chunkEDY72VWLjs.fromHex; exports.getBuyURL = _chunkEDY72VWLjs.getBuyURL; exports.gweiDecToWEIBN = _chunkEDY72VWLjs.gweiDecToWEIBN; exports.handleFetch = _chunkEDY72VWLjs.handleFetch; exports.hexToBN = _chunkEDY72VWLjs.hexToBN; exports.hexToText = _chunkEDY72VWLjs.hexToText; exports.isInfuraNetworkType = _chunkLGGBBU6Gjs.isInfuraNetworkType; exports.isNetworkType = _chunkLGGBBU6Gjs.isNetworkType; exports.isNonEmptyArray = _chunkEDY72VWLjs.isNonEmptyArray; exports.isPlainObject = _chunkEDY72VWLjs.isPlainObject; exports.isSafeChainId = _chunkEDY72VWLjs.isSafeChainId; exports.isSafeDynamicKey = _chunkEDY72VWLjs.isSafeDynamicKey; exports.isSmartContractCode = _chunkEDY72VWLjs.isSmartContractCode; exports.isValidHexAddress = _chunkEDY72VWLjs.isValidHexAddress; exports.isValidJson = _chunkEDY72VWLjs.isValidJson; exports.isValidSIWEOrigin = _chunkM6G3ISXGjs.isValidSIWEOrigin; exports.normalizeEnsName = _chunkEDY72VWLjs.normalizeEnsName; exports.parseDomainParts = _chunkM6G3ISXGjs.parseDomainParts; exports.query = _chunkEDY72VWLjs.query; exports.safelyExecute = _chunkEDY72VWLjs.safelyExecute; exports.safelyExecuteWithTimeout = _chunkEDY72VWLjs.safelyExecuteWithTimeout; exports.successfulFetch = _chunkEDY72VWLjs.successfulFetch; exports.timeoutFetch = _chunkEDY72VWLjs.timeoutFetch; exports.toChecksumHexAddress = _chunkEDY72VWLjs.toChecksumHexAddress; exports.toHex = _chunkEDY72VWLjs.toHex; exports.weiHexToGweiDec = _chunkEDY72VWLjs.weiHexToGweiDec;
