// @metamask/snaps-rpc-methods/dist/cjs/permitted/handlers.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    methodHandlers: function() {
        return methodHandlers;
    },
    handlers: function() {
        return handlers;
    }
});
const _getAllSnaps = require("./getAllSnaps");
const _getClientStatus = require("./getClientStatus");
const _getFile = require("./getFile");
const _getSnaps = require("./getSnaps");
const _invokeKeyring = require("./invokeKeyring");
const _invokeSnapSugar = require("./invokeSnapSugar");
const _requestSnaps = require("./requestSnaps");
const methodHandlers = {
    wallet_getAllSnaps: _getAllSnaps.getAllSnapsHandler,
    wallet_getSnaps: _getSnaps.getSnapsHandler,
    wallet_requestSnaps: _requestSnaps.requestSnapsHandler,
    wallet_invokeSnap: _invokeSnapSugar.invokeSnapSugarHandler,
    wallet_invokeKeyring: _invokeKeyring.invokeKeyringHandler,
    snap_getClientStatus: _getClientStatus.getClientStatusHandler,
    snap_getFile: _getFile.getFileHandler
};
const handlers = Object.values(methodHandlers);
