// @metamask/controller-utils/dist/util.mjs
import {
  BNToHex,
  PROTOTYPE_POLLUTION_BLOCKLIST,
  convertHexToDecimal,
  fetchWithErrorHandling,
  fractionBN,
  fromHex,
  getBuyURL,
  gweiDecToWEIBN,
  handleFetch,
  hexToBN,
  hexToText,
  isNonEmptyArray,
  isPlainObject,
  isSafeChainId,
  isSafeDynamicKey,
  isSmartContractCode,
  isValidHexAddress,
  isValidJson,
  normalizeEnsName,
  query,
  safelyExecute,
  safelyExecuteWithTimeout,
  successfulFetch,
  timeoutFetch,
  toChecksumHexAddress,
  toHex,
  weiHexToGweiDec
} from "./chunk-UFUF4N7T.mjs";
import "./chunk-XOU34P3V.mjs";
import "./chunk-7IUO3U3H.mjs";
export {
  BNToHex,
  PROTOTYPE_POLLUTION_BLOCKLIST,
  convertHexToDecimal,
  fetchWithErrorHandling,
  fractionBN,
  fromHex,
  getBuyURL,
  gweiDecToWEIBN,
  handleFetch,
  hexToBN,
  hexToText,
  isNonEmptyArray,
  isPlainObject,
  isSafeChainId,
  isSafeDynamicKey,
  isSmartContractCode,
  isValidHexAddress,
  isValidJson,
  normalizeEnsName,
  query,
  safelyExecute,
  safelyExecuteWithTimeout,
  successfulFetch,
  timeoutFetch,
  toChecksumHexAddress,
  toHex,
  weiHexToGweiDec
};
