// @metamask/controller-utils/dist/constants.mjs
import {
  ASSET_TYPES,
  ApprovalType,
  BUILT_IN_NETWORKS,
  CHAIN_ID_TO_ETHERS_NETWORK_NAME_MAP,
  ERC1155,
  ERC1155_INTERFACE_ID,
  ERC1155_METADATA_URI_INTERFACE_ID,
  ERC1155_TOKEN_RECEIVER_INTERFACE_ID,
  ERC20,
  ERC721,
  ERC721_ENUMERABLE_INTERFACE_ID,
  ERC721_INTERFACE_ID,
  ERC721_METADATA_INTERFACE_ID,
  FALL_BACK_VS_CURRENCY,
  GANACHE_CHAIN_ID,
  GWEI,
  IPFS_DEFAULT_GATEWAY_URL,
  MAX_SAFE_CHAIN_ID,
  NFT_API_BASE_URL,
  OPENSEA_PROXY_URL,
  ORIGIN_METAMASK,
  RPC,
  TESTNET_TICKER_SYMBOLS
} from "./chunk-XOU34P3V.mjs";
import "./chunk-7IUO3U3H.mjs";
export {
  ASSET_TYPES,
  ApprovalType,
  BUILT_IN_NETWORKS,
  CHAIN_ID_TO_ETHERS_NETWORK_NAME_MAP,
  ERC1155,
  ERC1155_INTERFACE_ID,
  ERC1155_METADATA_URI_INTERFACE_ID,
  ERC1155_TOKEN_RECEIVER_INTERFACE_ID,
  ERC20,
  ERC721,
  ERC721_ENUMERABLE_INTERFACE_ID,
  ERC721_INTERFACE_ID,
  ERC721_METADATA_INTERFACE_ID,
  FALL_BACK_VS_CURRENCY,
  GANACHE_CHAIN_ID,
  GWEI,
  IPFS_DEFAULT_GATEWAY_URL,
  MAX_SAFE_CHAIN_ID,
  NFT_API_BASE_URL,
  OPENSEA_PROXY_URL,
  ORIGIN_METAMASK,
  RPC,
  TESTNET_TICKER_SYMBOLS
};
